net/speedtest/speedtest_client.go

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package speedtest
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"time"
+)
+
+// RunClient dials the given address and starts a speedtest.
+// It returns any errors that come up in the tests.
+// It returns an error if the given test type isn't either download or upload.
+// If there are no errors in the test, it returns a slice of results.
+func RunClient(duration time.Duration, host, port string) ([]Result, error) {
+	c, err := net.Dial("tcp", net.JoinHostPort(host, port))
+	if err != nil {
+		return nil, err
+	}
+
+	config := TestConfig{TestDuration: duration, Version: version}
+
+	conn := c.(*net.TCPConn)
+	defer conn.Close()
+	encoder := json.NewEncoder(conn)
+
+	if err = encoder.Encode(config); err != nil {
+		return nil, err
+	}
+	var response TestConfigResponse
+	decoder := json.NewDecoder(conn)
+	if err = decoder.Decode(&response); err != nil {
+		return nil, err
+	}
+	if response.Error != "" {
+		return nil, errors.New(response.Error)
+	}
+	return runTestC(conn, config)
+}
+
+// runTestC handles the entire download speed test.
+// It has a loop that breaks if the connection recieves an IO error or if the server sends a header
```

Comment:
As written, this will not take into account the initial warm-up time required to establish a direct connection. The results might be hard to interpret. To my mind, the whole value to having a built-in tailscale speed test (instead of using iperf) is that we can be aware of Tailscale-specific quirks, like DERP vs direct.

Maybe that means having a "-direct" flag that only starts the speed test if/when a direct connection is established, and exits if the direct connection disappears.

Maybe that means for each interval printing whether data for that interval was transported direct or using DERP or both. (I am drawn to this.)

Reply:
Could we implement this when speedtest is added to tailscaled? Currently the speedtest code doesn't use any Tailscale code.

Reply:
OK. Please leave a TODO.